darwin.libcxx: (╯°□°）╯︵ ┻━┻

reckenrode opened the pull request:
Darwin currently uses libc++ from nixpkgs, but it probably shouldn’t (at least not by default). The libc++ that is built in nixpkgs comes from LLVM. [According][1] to Louis Dionne (of Apple and a member of the C++ committee), the libc++ shipped by Apple should be considered distinct from the libc++ in LLVM. We’ve been lucky that we have not had problems with ODR violations. This PR was prompted by leakage of symbols from the system libc++ that started happening with the release of macOS 15.4. It changes a number of things to improve the compatibility of C++ on Darwin.

* It makes the system libc++ the default C++. The Darwin bootstrap is built against it. The default stdenv uses it. This is done by making the libc++ headers and text-based stubs available from the 15.2 SDK. This is safe to do because using symbols that are not available on your deployment target results in a compile-time error.

### Testing

I also confirmed that all of the packages updated in this PR built (except for servo, which is broken due to Darwin having a case-insensitive build location), that Qt 5 and opencv4 built on x86_64-darwin, and that my [configs](https://github.com/reckenrode/nixos-configs) built (except for util-linux, which appears broken again on Darwin).

## Breaking Change

The primary points of breakage are as follows:

* Packages (such as Bazel and MoltenVK) that work around [#150655](https://github.com/NixOS/nixpkgs/issues/150655) by referencing `libcxx` should instead reference `stdenv.cc.libcxx`. (Those are also fixed in this PR.)
* The system libc++ expects Clang 16 or newer (due to libc++ support policies). To build with an older Clang, you need to override the SDK used by libc++ to make it use older libc++ headers.
* `overrideLibcxx` doesn’t work with the system libc++. If you really need to use it, you should use it with `llvmPackages.libcxxStdenv`.

# Notes about fixed packages

* Several packages depend on qt6.qtdeclarative, but it doesn’t build. Fixed in [#398849](https://github.com/NixOS/nixpkgs/pull/398849).
* python3Packages.py3exid2 looks for Boost in `/usr/local/lib`. Fixed in [#399005](https://github.com/NixOS/nixpkgs/pull/399005).

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc
[1]: https://discourse.llvm.org/t/apples-libc-now-provides-std-type-descriptor-t-functionality-not-found-in-upstream-libc/73881/3

reckenrode committed:
doc/stdenv/platform-notes: add section about libc++

reckenrode committed:
doc/rl-2511: mention libc++ changes on Darwin

reckenrode committed:
llvmPackages.clang: add `systemLibcxxClang` (primarily for Darwin)

reckenrode committed:
libtorch-bin: drop libc++ replacement

Darwin now uses the system libc++ by default, so drop the replacement.
The nixpkgs libc++ is not ABI-compatible with the system libc++, so this
replacement was not guaranteed to work anyway.

See: https://discourse.llvm.org/t/apples-libc-now-provides-std-type-descriptor-t-functionality-not-found-in-upstream-libc/73881/3

reckenrode committed:
ovftool: drop libc++ replacement

The libc++ in nixpkgs is not ABI-compatible with the system libc++, so
replacements are not guaranteed to work.

See: https://discourse.llvm.org/t/apples-libc-now-provides-std-type-descriptor-t-functionality-not-found-in-upstream-libc/73881/3

reckenrode committed:
pkgs/stdenv/darwin: switch to using the system libc++ by default

reckenrode committed:
darwin.libcxx: init at 19.1.2+apple-sdk-15.5

Adds the headers and text-based stubs for the system libc++.

This package is using the newest available version (currently 15.5)
because libc++ uses static availability checks to prevent using features
not supported by older deployment targets.

Co-authored-by: Emily <vcs@emily.moe>

RossComputerGuy commented on pkgs/development/compilers/llvm/common/libcxx/default.nix:
```diff
@@ -134,52 +165,73 @@ let
+    lib.optionals stdenv.hostPlatform.isDarwin (
+      # Add libc++abi reexports logic to `HandleLibCXXABI.cmake` from libcxxabi. It’s needed by libc++ to know which
+      # symbols to reexport from libc++abi (since not all of them should be).
+      lib.optional (lib.versionOlder release_version "15") (substitute {
+        src = ./libcxx-add-libcxxabi-reexports-llvm12.patch;
+        substitutions = [
+          "--subst-var-by"
+          "libcxxabi-build-support"
+          darwin.libcxxabi.build-support
+        ];
+      })
+      ++ lib.optional (lib.versionAtLeast release_version "15") (substitute {
+        src = ./libcxx-add-libcxxabi-reexports.patch;
+        substitutions = [
+          "--subst-var-by"
+          "libcxxabi-build-support"
+          darwin.libcxxabi.build-support
+        ];
+      })
+      # libc++ 19+ do not support reexporting symbols from the system libc++abi, but support can be trivially restored.
+      ++ lib.optional (lib.versionAtLeast release_version "19") ./libcxx-reexport-libcxxabi-symbols.patch
```
I think this is where `getVersionFile` can be useful.

RossComputerGuy commented on pkgs/development/compilers/llvm/common/default.nix:
```diff
@@ -218,6 +220,13 @@ let
         extraBuildCommands = mkExtraBuildCommands cc;
       };
 
+      systemLibcxxClang = wrapCCWith rec {
+        cc = tools.clang-unwrapped;
+        libcxx = darwin.libcxx;
+        extraPackages = [ targetLlvmLibraries.compiler-rt ];
+        extraBuildCommands = mkExtraBuildCommands cc;
+      };
+
```
Let's add a comment here to explain what this is useful for. I plan on documenting LLVM in nixpkgs so having a comment for reference will be helpful.

reckenrode replied:
Sure. It would be cool if this were something that could be controlled by https://github.com/NixOS/nixpkgs/pull/365057.

RossComputerGuy replied:
It may be, the logic won't be implemented in that PR. However, that PR does form the basis. Another possibility is when I refactor `pkgs/development/compilers/llvm/common/default.nix`, a special function could be provided to easily create a LLVM/Clang CC wrapper.

reckenrode replied:
With all the different C++ and C++ ABI libraries, it would be nice if the boilerplate could be reduced and switching between them made easier.

reckenrode committed:
pkgs/stdenv/cross: use the system libc++ by default on Darwin

reckenrode commented:
Fixed the signatures. I think I’m going to need to look into Sequoia as a replacement for GnuPG ….

Anyway, I confirmed that all of the packages changed in this PR build except for servo, which fails due to having `README.md` and `Readme.md` in one of its vendored crates (i.e., case-insensitive `/tmp` issues). I also confirmed that both Qt 5 and opencv4 build with their workarounds reverted from the last staging-next cycle.

reckenrode committed:
Revert "apple-sdk: add hack for C++ issues on `x86_64-darwin`"

This reverts commit 04b49f4e8d44ccd9f13b58ac4c31f8af4c28d4a1.

This workaround has a non-trivial cost in build time. Since it’s not
needed when linking against the system libc++, just drop it.

alyssais reviewed (commented):
The LLVM patches are unfortunate — could we not have upstream provide a way to use the system libcxx?

reckenrode commented on pkgs/development/compilers/llvm/common/compiler-rt/default.nix:
```diff
@@ -255,6 +255,13 @@ stdenv.mkDerivation (finalAttrs: {
+    + lib.optionalString (stdenv.hostPlatform.isDarwin && !noSanitizers) ''
+      substituteInPlace lib/tsan/rtl/tsan_interceptors_mac.cpp \
+        --replace-fail 3__1 ${toString (lib.stringLength stdenv.cc.libcxx.abiNamespace)}${stdenv.cc.libcxx.abiNamespace}
+    ''
```
I’m not sure it’s right. If you replace the stdenv’s libc++ without rebuilding clang (and compiler-rt), it’s not going to have an updated TSan. Might it be better to drop it and just disable TSan when building compiler-rt against a custom ABI namespace?

emilazy commented:
My thinking was that it seems like we’re pretty set on using the system `libc++` for Darwin, so it’s less clear if we have a use for a Nix‐built `libc++` to justify the maintenance load of the LLVM patching required, so it might make sense to focus this PR on the former and leave thinking about the latter to later?

reckenrode commented:
> My thinking was that it seems like we’re pretty set on using the system `libc++` for Darwin, so it’s less clear if we have a use for a Nix‐built `libc++` to justify the maintenance load of the LLVM patching required, so it might make sense to focus this PR on the former and leave thinking about the latter to later?

Essentially, drop the system libc++abi patches and let the `llvmPackages.libcxx` ODR violations continue since they’ve been mostly harmless so far?

reckenrode commented:
Should the top-level libcxx be the Darwin one instead of the LLVM one?

reckenrode commented:
There’s still an open question about the top-level `libcxx`, but I updated the PR to drop the LLVM changes. It now only switches Darwin to the system libc++. I may open a PR for the `llvmPackage.libcxx` conditional to simplify it, but it’s not really a big deal as-is currently.

reckenrode commented on pkgs/by-name/ap/apple-sdk/common/process-stubs.nix:
```diff
@@ -38,21 +38,5 @@ self: super: {
-        ' > $stub~
-        llvm-readtapi --filetype=tbd-v4 $stub~ -o $stub
-        rm $stub~
-      done < <(find . -name '*.tbd' -print0)
```
I’ve updated the commit message to indicate why the workaround is being dropped.

philiptaron commented:
Randy, what testing if any do you need?

alyssais reviewed (commented):
Ack.

Ericson2314 commented:
This seems good and minimal now.

Now that I have made top-level `libc` be good, I think it's fair to say we can make a top-level `libcxx` that works in much the same way.

In particular, a `libcxx = string` in stdenv hostPlatform will get us towards @RossComputerGuy's PR (the tool changes part of that is harder, but the library parts are good)

emilazy commented on doc/release-notes/rl-2511.section.md:
```diff
@@ -9,6 +9,10 @@
   supported. `pkgsCross.loongarch64-linux-embedded` can be used to build software and systems for these platforms.
 
+- Darwin has switched to using the system libc++. This was done for improved compatibility and to avoid ODR violations.
+  If a newer C++ library feature is not available on the default deployment target, you will need to increase the deployment target or use libc++ from LLVM.
```
Maybe shouldn’t recommend the LLVM `libc++` at present? And this is also probably zero impact for most users, so it might belong in the backward incompatibilities section instead.

reckenrode replied:
Yeah, might just keep the deployment target note and also update the Darwin documentation not to mention it.

reckenrode replied:
I updated the Darwin documentation to mention that you can use libc++ from LLVM, but it’s not recommended.

emilazy commented on pkgs/os-specific/darwin/libcxx/default.nix:
```diff
@@ -0,0 +1,26 @@
+    runHook postInstall
+  '';
+
+  passthru = {
+    abiNamespace = "__1";
+    isLLVM = true;
+  };
```
These are currently unused, right?

reckenrode replied:
They’re leftover from the libc++abi patchset. I can drop this part.

reckenrode replied:
The `abiNamespace` was unused, but `isLLVM` is used by `cc-wrapper` to determine if the C++ library is from LLVM or not. Dropping it breaks the wrapper, so I only dropped the ABI namespace.

emilazy commented on pkgs/os-specific/darwin/libcxx/default.nix:
```diff
@@ -0,0 +1,26 @@
+
+stdenvNoCC.mkDerivation (finalAttrs: {
+  pname = "libcxx";
+  inherit (apple-sdk) src version;
```
I’m a bit worried that macOS SDK versions look enough like LLVM versions that people might be smart and try to condition on `libc++` version and fail. But I’m not sure if we can do anything better, either…

reckenrode replied:
I don’t know what to put. There’s enough misunderstanding around what version of Clang is shipped by Apple that people might reflexively go with the LLVM one because it is slightly newer, or they might be confused when 14.x has an actually older runtime (or 26.0 has a newer one).

Are there any examples of conditioning on libc++ version?

emilazy replied:
`overrideLibcxx` in `pkgs/stdenv/adapters.nix` does, I wonder who we should blame for that? 😆

But since Swift and Qt 5 are the only user (and hopefully Swift won’t be soon? and the Qt 5 one seems dubious to me), that’s hopefully not the end of the world; we can potentially just drop it and the assertion is harmless anyway since 15 > 12.

This is not a blocker for me, but it would be nice for the package name in the store path to be distinguished from an LLVM one, and maybe we can solve both problems at once. Some proposals:

* `libcxx-apple-sdk-15.2`
* `libcxx-0-apple-sdk-15.2` (ensures version comparisons will at least fail harder)

I think that *ideally*, we’d have something like `libcxx-⟨closest upstream LLVM version⟩-apple-sdk-⟨macOS SDK version⟩`. But it might be hard to reliably determine the corresponding LLVM version, given what Apple’s tree is like.

emilazy reviewed (commented):
Idea agreed upon a while ago, almost all LGTM, some comments around implementation and messaging.

emilazy reviewed (approved):
Needs rebase; otherwise LGTM. Left a comment about the one unresolved thing (version).

emilazy committed:
apple-sdk_15: 15.2 -> 15.5

emilazy committed:
overrideLibcxx: drop

We now use the latest SDK’s libc++ headers, and longer have Clang
versions old enough for version compatibility to be a concern.
